contracts/BonusRewards.sol

```diff
@@ -3,11 +3,363 @@
+      address _lpToken = _lpTokens[i];
+      User memory user = users[_lpToken][msg.sender];
+      if (user.amount == 0) continue;
+      _updatePool(_lpToken);
+      _claimRewards(_lpToken, user);
+      _updateUserWriteoffs(_lpToken);
+    }
+  }
+
+  function deposit(address _lpToken, uint256 _amount) external override nonReentrant notPaused {
+    require(pools[_lpToken].lastUpdatedAt > 0, "Blacksmith: pool does not exists");
+    require(IERC20(_lpToken).balanceOf(msg.sender) >= _amount, "Blacksmith: insufficient balance");
+
+    _updatePool(_lpToken);
+    User storage user = users[_lpToken][msg.sender];
+    _claimRewards(_lpToken, user);
+    user.amount = user.amount + _amount;
+    _updateUserWriteoffs(_lpToken);
+
+    IERC20(_lpToken).safeTransferFrom(msg.sender, address(this), _amount);
+    emit Deposit(msg.sender, _lpToken, _amount);
+  }
+
+  /// @notice withdraw up to all user deposited
+  function withdraw(address _lpToken, uint256 _amount) external override nonReentrant notPaused {
+    require(pools[_lpToken].lastUpdatedAt > 0, "Blacksmith: pool does not exists");
+    _updatePool(_lpToken);
+
+    User storage user = users[_lpToken][msg.sender];
+    _claimRewards(_lpToken, user);
+    uint256 amount = user.amount > _amount ? _amount : user.amount;
+    user.amount = user.amount - amount;
+    _updateUserWriteoffs(_lpToken);
+
+    _safeTransfer(_lpToken, amount);
+    emit Withdraw(msg.sender, _lpToken, amount);
+  }
+
+  /// @notice withdraw all without rewards
+  function emergencyWithdraw(address _lpToken) external override nonReentrant {
+    User storage user = users[_lpToken][msg.sender];
+    uint256 amount = user.amount;
+    user.amount = 0;
+    _safeTransfer(_lpToken, amount);
+    emit Withdraw(msg.sender, _lpToken, amount);
+  }
+
+  /// @notice called by authorizers only
+  function addBonus(
+    address _lpToken,
+    address _bonusTokenAddr,
+    uint256 _startTime,
+    uint256 _weeklyRewards,
+    uint256 _transferAmount
+  ) external override nonReentrant notPaused {
+    require(_isAuthorized(msg.sender, allowedTokenAuthorizers[_lpToken][_bonusTokenAddr]), "BonusRewards: not authorized caller");
+    require(_startTime >= block.timestamp, "BonusRewards: startTime in the past");
+
+    // make sure the pool is in the right state (exist with no active bonus at the moment) to add new bonus tokens
+    Pool memory pool = pools[_lpToken];
+    require(pool.lastUpdatedAt != 0, "BonusRewards: pool does not exist");
+    Bonus[] memory bonuses = pool.bonuses;
+    for (uint256 i = 0; i < bonuses.length; i++) {
+      if (bonuses[i].bonusTokenAddr == _bonusTokenAddr) {
+        // when there is alreay a bonus program with the same bonus token, make sure the program has ended properly
+        require(bonuses[i].endTime + WEEK < block.timestamp, "BonusRewards: last bonus period hasn't ended");
+        require(bonuses[i].remBonus == 0, "BonusRewards: last bonus not all claimed");
```

Comment:
Does this mean we can't add a bonus without it being totally cleared out by us or by users?

Reply:
Yes.